Add torch.dot for complex tensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42745 Add torch.dot for complex tensors**

Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

anjali411 committed:
Add dot product for complex tensors

[ghstack-poisoned]

anjali411 committed:
Update on "Add dot product for complex tensors"


TODO: potentially add a fast path for complex dot

[ghstack-poisoned]

anjali411 committed:
Update on "Add dot product for complex tensors"


TODO: potentially add a fast path for complex dot

[ghstack-poisoned]

anjali411 committed:
Update on "Add dot product for complex tensors"


TODO: potentially add a fast path for complex dot

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -6,6 +6,8 @@
 
 #if AT_BUILD_WITH_BLAS()
 extern "C" double ddot_(int *n, double *x, int *incx, double *y, int *incy);
+extern "C" void zdotu_(std::complex<double> *res, int *n, std::complex<double> *x, int *incx, std::complex<double> *y, int *incy);
```
https://www.math.utah.edu/software/c-with-fortran.html#function-return-types

"...you should not expect to use Fortran functions that return types such as COMPLEX or COMPLEX*16. Write a SUBROUTINE interface to your Fortran function instead, and then invoke it as a void function from C or C++."

anjali411 committed:
Update on "Add torch.dot for complex tensors"



[ghstack-poisoned]

ezyang commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -490,6 +490,20 @@ void dot<float>(CUDABLAS_DOT_ARGTYPES(float)) {
 
+template <>
+void dot<c10::complex<double>>(CUDABLAS_DOT_ARGTYPES(c10::complex<double>)) {
+  TORCH_CUDABLAS_CHECK(cublasZdotu(handle, n, reinterpret_cast<const cuDoubleComplex*>(x),
```
Hmm, I wonder if we shouldn't have some methods on `c10::complex` for doing pointery conversions like this. It would be nice to not have to be slinging reinterpret cast everywhere. (No action needed for PR)

zasdfgbnm commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -490,6 +490,20 @@ void dot<float>(CUDABLAS_DOT_ARGTYPES(float)) {
 }
 
+template <>
+void dot<c10::complex<double>>(CUDABLAS_DOT_ARGTYPES(c10::complex<double>)) {
```
Looks like you forget to add the template specification to `CUDABlas.h`?

anjali411 replied:
yeah that did fix the windows test failures. thanks!!

anjali411 committed:
Update on "Add torch.dot for complex tensors"



[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.dot for complex tensors"


Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.dot for complex tensors"


Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.dot for complex tensors"


Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.dot for complex tensors"


Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.dot for complex tensors"


Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

[ghstack-poisoned]

jeffdaily commented:
@anjali411 as discussed offline, the reason CPU results for the ROCm build are failing is because `zdotu_` and `cdotu_` symbols from our openblas install are getting linked and called instead of the static inline versions that wrap the cblas calls.  Suggested change:

```diff
diff --git a/aten/src/ATen/native/BlasKernel.cpp b/aten/src/ATen/native/BlasKernel.cpp
index ef05cb8..1fe8a73 100644
--- a/aten/src/ATen/native/BlasKernel.cpp
+++ b/aten/src/ATen/native/BlasKernel.cpp
@@ -17,9 +17,6 @@ extern "C" void sgemv_(char *trans, int *m, int *n, float *alpha, float *a, int
 # define ffloat float
 #endif

-extern "C" ffloat sdot_(int *n, float *x, int *incx, float *y, int *incy);
-extern "C" void cdotu_(std::complex<float> *res, int *n, std::complex<float> *x, int *incx, std::complex<float> *y, int *incy);
-extern "C" void zdotu_(std::complex<double> *res, int *n, std::complex<double> *x, int *incx, std::complex<double> *y, int *incy);

 #ifdef BLAS_USE_CBLAS_DOT
 extern "C" float cblas_sdot(const int n, const float *x, const int incx, const float *y, const int incy);
@@ -40,6 +37,10 @@ static inline void zdotu_(std::complex<double> *res, const int *n, const std::co
   cblas_zdotu_sub(*n, x, *incx, y, *incy, res);
 }
 #endif // THBlas_cblas_dot_
+#else // BLAS_USE_CBLAS_DOT
+extern "C" ffloat sdot_(int *n, float *x, int *incx, float *y, int *incy);
+extern "C" void cdotu_(std::complex<float> *res, int *n, std::complex<float> *x, int *incx, std::complex<float> *y, int *incy);
+extern "C" void zdotu_(std::complex<double> *res, int *n, std::complex<double> *x, int *incx, std::complex<double> *y, int *incy);
 #endif // BLAS_USE_CBLAS_DOT
 #endif // AT_BUILD_WITH_BLAS
```

jeffdaily commented:
BLAS development began in fortran, and the function calling convention differs between compilers and compiler flags, making it difficult to get correct when calling the fortran function from C.  Sometimes the complex value is returned as a hidden positional argument as in your extern declaration, and sometimes it is returned like a regular C function call return.  Since the openblas symbols were getting linked, and their extern declaration was incorrect, the function was getting called but there was a mismatch between the expected function signature and how it was declared.

anjali411 committed:
Update on "Add torch.dot for complex tensors"


Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

[ghstack-poisoned]

anjali411 committed:
Update on "Add torch.dot for complex tensors"


Differential Revision: [D23056382](https://our.internmc.facebook.com/intern/diff/D23056382)

[ghstack-poisoned]

gchanan commented on aten/src/TH/generic/THBlas.cpp:
```diff
@@ -17,12 +17,21 @@ TH_EXTERNC void saxpy_(int *n, float *a, float *x, int *incx, float *y, int *inc
+TH_EXTERNC void cblas_cdotu_sub(const int n, const void *x, const int incx, const void *y, const int incy, void *dotu);
+TH_EXTERNC void cblas_zdotu_sub(const int n, const void *x, const int incx, const void *y, const int incy, void *dotu);
+
+#ifndef THBlas_cblas_dot_
```
why are these symbols being defined?

anjali411 replied:
removed in https://github.com/pytorch/pytorch/pull/43148
